climakitae/data_loaders.py

```diff
@@ -0,0 +1,125 @@
+import xarray as xr
+from shapely.geometry import box #, Point, Polygon
+import regionmask
+import intake
+import numpy as np
+
+# support methods for core.Application.generate
+
+
+def _experiment_id(scenario):
+    """
+    Returns the SSP designation in CMIP 'experiment_id' format based on the
+    contents of the human-readable scenario designation stored in 'selections'.
+    """
+    split = scenario.split(" ")
+    for i, one in enumerate(split):
+        if "SSP" in one:
+            numbers = split[i + 1].split("-")
+            ssp = numbers[0]
+            rcp = numbers[1].split(".")
+            rcp = rcp[0] + rcp[1]
+            return "ssp" + ssp + rcp
+        elif one == "Historical Climate":
+            return "historical"
+        elif one == "Historical Reconstruction":
+            return ""
+    return None
+
+
+def _get_file_list(cat, selections, scenario):
+    """
+    Returns a list of simulation names for all of the simulations present in the catalog
+    for a given scenario, contingent on other user-supplied constraints in 'selections'.
+    """
+    file_list = []
+    for item in list(cat):
+        if cat[item].metadata["nominal_resolution"] == selections.resolution:
+            if cat[item].metadata["experiment_id"] == _experiment_id(scenario):
+                file_list.append(cat[item].name)
+    return file_list
+
+
+def _get_var_name(cat, var_description):
+    _ds = cat[list(cat)[0]].to_dask()
+    for i in _ds.data_vars:
+        if _ds[i].attrs["description"] == var_description:
+            return i
+    return "T2"  # a default that might not be what the user asked for...
+    # add some handling for if it's not there for some reason,
+    # to provide a more useful error message instead of returning whatever default
+
+
+def _open_and_concat(cat, file_list, selections, geom):
+    """
+    Open multiple zarr files, and add them to one big xarray Dataset. Coarsens in time, and/or
+    subsets in space if selections so-indicates. Won't work unless the file_list supplied
+    contains files of only one nominal resolution (_get_file_list ensures this).
+    """
+    all_files = xr.Dataset()
+    for one_file in file_list:
+        with cat[one_file].to_dask() as data:
+            source_id = data.attrs["source_id"]
+            if selections.variable == "precipitation (total)":
+                pass
+            elif selections.variable == "wind 10m magnitude":
+                pass
+            else:
+                var_name = _get_var_name(cat, selections.variable)
+                data = data[var_name]
+            # coarsen in time if 'selections' so-indicates:
+            if selections.timescale == "daily":
+                data = data.resample(time="1D").mean("time")
+            elif selections.timescale == "monthly":
+                data = data.resample(time="1MS").mean("time")
+            if geom:
+                # subset data spatially:
+                ds_region = regionmask.Regions(
```

Comment:
This could be handled with `rasterio.features.geometry_mask()` which is already installed on our docker image, so that could avoid adding a new dependency to `regionmask` just for spatial subsetting which might be nice.

Reply:
so, I see your point about not adding another dependency. I was messing around with rasterio, but I couldn't get it to work, and this does work. Perhaps instead I make a card to go back and try to do that later?

Reply:
Good idea, feel free to assign me to the card and I can work on that.